system/mp/mpi_wrapper.cpp

```diff
@@ -0,0 +1,526 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+#include <cstdlib>
+#include <queue>
+#include <typeinfo>
+#include "mpi_wrapper.hpp"
+#include "mputil.hpp"
+
+#define WAIT_DELAY 100
+#define PROBING_DELAY 100
+
+//----------Functions and Data structures managing communication data related to Send/Recv Communications in orogress-----------//
+
+// Data structure to keep the data relating to send/receive communications
+class CommData
+{
+private:
+    bool send;                      // TRUE => relates to a send communication | FALSE => relates to receive communication
+    bool locked = false;
+    bool cancellationLock = false;
+    bool cancellationInProgress = false;
+    CriticalSection dataChangeLock;
+    void lock(){dataChangeLock.enter(); locked=true;}
+    void unlock(){locked=false; dataChangeLock.leave();}
+public:    
+    bool isSend(){return send;}
+    bool isReceive(){return !isSend();}
+    bool isEqual(bool _send, int _rank, int _tag)
+    {
+        return (send==_send) && (_rank==MPI_ANY_SOURCE || rank==_rank) && (_tag==MPI_ANY_TAG || tag==_tag);
+    }
+
+    int rank;                       // source/destination rank of the processor
+    int tag;                        // MPI tag information
+    MPI_Request* request;           // persistent request object to keep track of ongoing MPI call
+    const hpcc_mpi::MPIComm& comm;                // MPI communicator
+
+
+    CommData(bool _send, int _rank, int _tag, const hpcc_mpi::MPIComm& _comm):
+        send(_send), rank(_rank), tag(_tag), comm(_comm), request(new MPI_Request()){}
+
+    CommData(int _rank, int _tag, const hpcc_mpi::MPIComm& _comm): rank(_rank), tag(_tag), comm(_comm), send(false), request(new MPI_Request()){}
+
+    void notifyCancellation()
+    {
+        lock();
```

Comment:
I didn't want to use it in this case because I wanted to keep track of the locking/unlocking in those separate functions. We currently have 0 test cases to test cancellation. And once we have fully tested it, we can change it to use CriticalBlock.